Add utility upload function

eba99 opened the pull request:
New classes created for uploads to v1 and v2 lake, as well as modifying read_and_upload_file_to_dash function to do the following: 
- Determine which version a client is on (can also be specified)
- Upload file to the correct endpoint

Effect is that users should not have to make changes to their uploads, unless they are using an application client and secret.  These must be passed as args in the function to retrieve upload credentials.

eba99 committed:
Update dash.py

eba99 committed:
Finalize first draft of utility function to upload to v2 lake

eba99 committed:
Create virtual environment

eba99 committed:
Enable modify_lambda for v2 upload

eba99 committed:
Add handling of dryrun to v2 upload

eba99 committed:
Allow forcing of data model version

eba99 committed:
Update dash.py

eba99 committed:
Update dash.py

eba99 committed:
Update dash.py

eba99 committed:
Update dash.py

eba99 committed:
Update dash.py

eba99 committed:
Write a utility function for df upload and modify general structure

eba99 committed:
Add default checksum on row count

eba99 committed:
Update dash.py

eba99 committed:
Merge branch 'add_cleaning_up_status' into add_utility_upload_function

eba99 committed:
Update dash.py

eba99 committed:
Merge branch 'main' into add_utility_upload_function

eba99 committed:
Set commit_after_upload default to True

eba99 committed:
Install dotenv for virtual environment

eba99 committed:
Add ability to upload from application

eba99 committed:
Validate application config parameter

eba99 committed:
Add application config handling to v2 easy upload functions

timalive commented on Pipfile:
```diff
@@ -15,6 +15,8 @@ tqdm = ">=4.64.0"
 aiohttp = "*"
 pydantic = ">2"
 awswrangler = {extras = ["all"], version = ">=3.4.2"}
+comotion-sdk = {file = ".", editable = true}
```
Cant have this here, as it creates circular references.  Having it in the dev package is fine for local testing.

timalive replied:
@eba99 this still needs to be resolved

eba99 replied:
Has been removed and will be tested with new pipfile

eba99 replied:
Ok should now be removed

timalive commented on Pipfile:
```diff
@@ -15,6 +15,8 @@ tqdm = ">=4.64.0"
 pydantic = ">2"
 awswrangler = {extras = ["all"], version = ">=3.4.2"}
+comotion-sdk = {file = ".", editable = true}
+python-dotenv = "*"
```
not sure why you require this?

timalive replied:
Still awaing response on this one

eba99 replied:
Has since been removed

timalive commented on src/comotion/dash.py:
```diff
@@ -215,6 +220,30 @@ def generate_presigned_url_for_file_upload(self, file_key: str = None) -> FileUp
 
         return self.load_api_instance.generate_presigned_url_for_file_upload(self.load_id, file_upload_request=file_upload_request)
         
+    def upload_file(self, file: Union[str, io.FileIO], file_upload_response: FileUploadResponse):
```
is there a unit test for this?

timalive replied:
awating answer

eba99 replied:
Unit test added in this commit, for this and upload_query as that was also missing: https://github.com/ComotionLabs/comotion-sdk/commit/40cdda6a79474e3aca7e0e29f011c4d5fa3604dc

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +455,263 @@ def stop(self):
+
+        return load
+    
+class Dash_v1_Easy_Upload():
```
why do you need a whole class for v1 and v2.  PErhaps combine and call it UploadUtilities?

timalive replied:
also, unit tests

timalive replied:
upload utilities

timalive replied:
still awaiting response on this

eba99 replied:
This has been removed and is only using the v1_upload_csv class.

timalive commented on src/comotion/dash.py:
```diff
@@ -511,12 +797,19 @@ def read_and_upload_file_to_dash(
+    data_model_version: str = None
 ):
     """Reads a file and uploads to dash.
 
```
add new parameters to comments

timalive replied:
why would you have to supply a load to this function?

timalive replied:
doc string

timalive replied:
see above

eba99 replied:
These arguments have since been removed. Deprecation warning added to v1_upload_csv and read_and_upload_to_dash: https://github.com/ComotionLabs/comotion-sdk/commit/4d91d978ce5d35e09cc0b95ac4352c552e28ee5a

timalive commented on src/comotion/dash.py:
```diff
@@ -567,46 +860,72 @@ def read_and_upload_file_to_dash(
+            # Get migration status
+            migration = Migration(config)
+            print('Migration Status: ' + str(migration.status().full_migration_status))
+            if migration.status().full_migration_status == 'Completed': # What about for new clients who start on v2 lake?  Will status be complete?
```
Good question. Until further notice, we must run a migration on all new clients as part of setting them up.

timalive commented on src/comotion/dash.py:
```diff
@@ -215,6 +220,30 @@ def generate_presigned_url_for_file_upload(self, file_key: str = None) -> FileUp
 
         return self.load_api_instance.generate_presigned_url_for_file_upload(self.load_id, file_upload_request=file_upload_request)
         
+    def upload_file(self, file: Union[str, io.FileIO], file_upload_response: FileUploadResponse):
```
add file_upload response as an optional. IF not supplied then automatically fetch.  Supply filekey as an optional.  Also, boolen option to use the file name as filekey.

timalive replied:
still to be resolved

eba99 replied:
Added with unit tests: https://github.com/ComotionLabs/comotion-sdk/commit/89c506d593e9c940cf210bf5400aac0dc4fee757

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +455,263 @@ def stop(self):
+        return load
+    
+class Dash_v1_Easy_Upload():
+    def v1_upload_csv(
```
doctring. v1 will be deprecated.

timalive replied:
also this

eba99 replied:
This has been resolved as these args are no longer in the function. Docstring improved.

eba99 committed:
comotion-sdk from packages in pipfile

timalive committed:
Update Pipfile.lock

eba99 committed:
Update pipfile and lock

eba99 committed:
Update Load.upload_file function

Make file_upload_response optional.  If not provided, create the response with optional file key.  Once file_upload_response exists, validation added.

eba99 committed:
Add exceptiongroup package to pipfile

eba99 committed:
Fix file reference in integration test for dash_upload_file

Use os.path.join instead of specifying full string

eba99 committed:
Add docstring to Load.upload_file()

eba99 committed:
Add upload tracker and dryrun path arg to Load class

eba99 committed:
Create add_upload function with append and overwrite options

eba99 committed:
Progress with DashBulkUploader

eba99 committed:
Write excute_uploads and remove_upload

eba99 committed:
Improve DashBulkUploader

Added add_load function
More flexible ability to handle different file types
Responses are stored within the uploader.uploads instead of having separate variables for upload status.

eba99 committed:
Fixed incorrect class variable call

eba99 committed:
Update docstrings

eba99 committed:
Update docstring for add_load function

eba99 committed:
Update docstrings

eba99 committed:
Update docstring for read_and_upload_file_to_dash function

eba99 committed:
Fix bugs in Load init and upload_file function

eba99 committed:
Fix bugs in add_data_to_load function

timalive commented on Pipfile:
```diff
@@ -15,6 +15,7 @@ tqdm = ">=4.64.0"
 aiohttp = "*"
 pydantic = ">2"
 awswrangler = {extras = ["all"], version = ">=3.4.2"}
+exceptiongroup = "*"
```
interesting. where are you using this

timalive commented on src/comotion/dash.py:
```diff
@@ -106,9 +115,28 @@ class Load():
     resulting load_id in `load_id`
 
-    If you wish to work with an existing load, then simply supply load_id parameter
+    If you wish to work with an existing load, then supply load_id only the load_id parameter
```
this sentence does not make sense. Probably wanted to say "then supply only the load_id parameter."?

timalive commented on src/comotion/dash.py:
```diff
@@ -106,9 +115,28 @@ class Load():
+    If you wish to work with an existing load, then supply load_id only the load_id parameter
+
+    Example of upload with a Load instance: 
+        load = Load(config = DashConfig(Auth('orgname')), 
```
we use "Sphynx" for documentation. that automatically converts these docstrings into proper docs.  This example should be a "literal block".  You can see how to do that [here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks).  Note that it is simply adding a double ::

timalive commented on src/comotion/dash.py:
```diff
@@ -136,8 +166,16 @@ def __init__(
             In the case where you want to work with an existing load on dash, supply this parameter, and no other parameter (other than config) will be required.
+        track_rows_uploaded: bool, optional
+            If True, track the number of rows uploaded with the current Load instance.  This can be used to automatically create a checksum on commit (see Load.commit), however is not recommended for 
+            large files as this may increase the duration of upload significantly.
```
nice

timalive commented on src/comotion/dash.py:
```diff
@@ -136,8 +166,16 @@ def __init__(
+            large files as this may increase the duration of upload significantly.
+        path_to_output_for_dryrun: str, optional
+            if specified, no upload will be made to dash, but files
+            will be saved to the location specified. This is useful for testing.
```
nice

timalive commented on src/comotion/dash.py:
```diff
@@ -136,8 +166,16 @@ def __init__(
         """
         load_data = locals()
+        lowerlevel_load_sig = signature(comodash_api_client_lowlevel.Load)
+        lowerlevel_load_keys = [key for key in lowerlevel_load_sig.parameters.keys()]
```
I learnt something

timalive commented on src/comotion/dash.py:
```diff
@@ -154,15 +192,26 @@ def __init__(
+                        key: value
+                        for key, value in load_data.items()
+                        if key in lowerlevel_load_keys
+                }
```
clever

timalive commented on src/comotion/dash.py:
```diff
@@ -215,7 +264,93 @@ def generate_presigned_url_for_file_upload(self, file_key: str = None) -> FileUp
+                file_key: str = None,
+                file_upload_response: FileUploadResponse = None):
+        """
+        Uploads a file to an S3 bucket using a presigned URL.
```
update this comment... should be file, or dataframe or directory depending on what is provided

timalive replied:
consider adding handling of a df iterator - lkike what reading a chunked CSV produces in panda.s

timalive commented on src/comotion/dash.py:
```diff
@@ -215,7 +264,93 @@ def generate_presigned_url_for_file_upload(self, file_key: str = None) -> FileUp
+
+        Parameters:
+        ----------
+        file : Union[str, io.FileIO]
```
should be "data"

timalive replied:
maybe change it to "data_source"

timalive commented on src/comotion/dash.py:
```diff
@@ -215,7 +264,93 @@ def generate_presigned_url_for_file_upload(self, file_key: str = None) -> FileUp
+        ----------
+        file : Union[str, io.FileIO]
+            The file to be uploaded. This can be a file path (str) or a file object (io.FileIO).
+        file_key : str, optional
```
is this only in the case of a file?

timalive replied:
saying "under which the file"  mayh be confusing for other formats

timalive commented on src/comotion/dash.py:
```diff
@@ -215,7 +264,93 @@ def generate_presigned_url_for_file_upload(self, file_key: str = None) -> FileUp
+            The file to be uploaded. This can be a file path (str) or a file object (io.FileIO).
+        file_key : str, optional
+            The key (path) under which the file will be stored in the S3 bucket. If not provided, it will be generated.
+        file_upload_response : FileUploadResponse, optional
```
will need more info here.   I am thinking we should kick it out into another function.  This will confuse python analysts

timalive commented on src/comotion/dash.py:
```diff
@@ -215,7 +264,93 @@ def generate_presigned_url_for_file_upload(self, file_key: str = None) -> FileUp
+
+        Example:
+        -------
+        >>> load = Load(dashconfig)
```
probably better to use a literal block

timalive commented on src/comotion/dash.py:
```diff
@@ -215,7 +264,93 @@ def generate_presigned_url_for_file_upload(self, file_key: str = None) -> FileUp
+            else:
+                print(f"Upload completed: {key}")
+
+            return upload_reponse
```
cant see where you are dealing with the upload a whole directory option?

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+    def __init__(self, 
+                 auth_token: Auth,
+                 default_chunksize: int = 30000) -> None:
+
```
needs a thorough docstring. Sholuld include a good overview of the class and how to use it.

timalive replied:
you can maybe include an example of using the class here

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+        ]
+
+    def refresh_dash_config(self):
+        """Return a new DashConfig object with the auth_token provided to DashBulkUploader."""
```
the name of this function not 100% cleaer I think

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+        
+        return start + middle + end
+    
+    def add_data_to_load(
```
how is this more functionality than what should be available in the load class?

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+    def execute_upload(
+        self,
+        table_name: str,
+        max_workers: int = 5
```
max workers should default to CPU's available x2 I think.
here is an idea of code for this:
```
if max_workers is None:
  num_cores = os.cpu_count()
  max_workers = num_cores * 2 
```

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+    
+    def execute_upload(
+        self,
+        table_name: str,
```
why does this have to be specific to one table? Cant we execute loads to a number of tables?

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+
+            upload_futures = []
+
+            with ThreadPoolExecutor(max_workers=max_workers) as upload_executor:  # Use multi-threading to speed up uploads
```
actually, I see this from the [threadpoolexecutor docs.](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor)
 if max_workers not supplied, we can just get threadpoolexecutor to figure out how many it should run.
`  Changed in version 3.5: If max_workers is None or not given, it will default to the number of processors on the machine, multiplied by 5, assuming that [ThreadPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor) is often used to overlap I/O instead of CPU work and the number of workers should be higher than the number of workers for [ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor).`

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+        table_names = [table_name for table_name in self.uploads.keys()]
+        self.execute_multiple_uploads(table_names = table_names)
+
+    def get_load_info(self, max_workers: int = 5):
```
see my previuos comment. we should remove the default.

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+            If an error occurs while retrieving the load information for any table, it is caught
+            and printed, but the function continues to retrieve the remaining load statuses.
+        """
+        with ThreadPoolExecutor(max_workers=max_workers) as status_executor:
```
what if we run into rate limit exceeded probelms on the API. we should have a quick chat about that scenario.

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+    def upload_df(
+        self,
+        df: pd.DataFrame,
+        load: Load,
```
why not just add this function to the Load class?

timalive commented on src/comotion/dash.py:
```diff
@@ -426,6 +571,613 @@ def stop(self):
+
+        return response
+
+    def upload_file(
```
why not just add this to the load class?

timalive commented on src/comotion/dash.py:
```diff
@@ -506,109 +1258,157 @@ def create_gzipped_csv_stream_from_df(df: pd.DataFrame) -> io.BytesIO:
+    dash_orgname : str
+        Orgname of your Dash instance.
+    file_key : str, optional
+        A unique key for the file being uploaded. If not provided, a key will be generated.
```
we should ad dmore explanation to this file_key comment.  Why do we have a file key? Something like "file key helps ensure that the same file is not uploaded twice"

timalive commented on src/comotion/dash.py:
```diff
@@ -506,109 +1258,157 @@ def create_gzipped_csv_stream_from_df(df: pd.DataFrame) -> io.BytesIO:
-    - upload them to dash
+    - Read a CSV file
+    - Break it up into multiple CSVs, each with a maximum number of lines defined by `chunksize`
+    - Upload them to Dash
```
add a general note here about how this will check if its v1 or v2.  And a link to an explanation of v1 and v2.

timalive commented on src/comotion/dash.py:
```diff
@@ -506,109 +1258,157 @@ def create_gzipped_csv_stream_from_df(df: pd.DataFrame) -> io.BytesIO:
+        This is useful for testing. Multiple files will be created (1 per chunk)
+    service_client_id : str, optional
+        If specified, specifies the service client for the upload. See the Dash documentation for an explanation
+        of the service client.
```
we can include a link to a page - we must have a proper explanation in the docs on service client and how it works with loads.

eba99 replied:
A link has since been added to docstring. Is this sufficient?

timalive commented on src/comotion/dash.py:
```diff
@@ -506,109 +1258,157 @@ def create_gzipped_csv_stream_from_df(df: pd.DataFrame) -> io.BytesIO:
+    load_type : str, default 'APPEND_ONLY'
+        The type of load operation. Default is 'APPEND_ONLY'.
+    data_model_version : str, optional
+        The data model version to use for the upload. If not specified, the function will determine the version.
```
we should say that this should only be used in exceptional circumstances.

eba99 replied:
This has been added

timalive commented on src/comotion/dash.py:
```diff
@@ -506,109 +1258,157 @@ def create_gzipped_csv_stream_from_df(df: pd.DataFrame) -> io.BytesIO:
+    data_model_version : str, optional
+        The data model version to use for the upload. If not specified, the function will determine the version.
+    entity_type : str, default Auth.USER
+        The entity type for authentication.
```
?

timalive reviewed (changes requested):
I think we should ask Kevin and Daniel to give us a few hours to test this.

eba99 committed:
Update based on comments

eba99 committed:
Move single upload functions to Load class

eba99 committed:
Add upload_dash_query function to Load class

eba99 committed:
Update create file key function and pipfile

eba99 committed:
Fix bugs in v1_upload_csv and Load init

eba99 committed:
Fix unit test after addition of extra parameters

eba99 committed:
Add TestDashBulkUploader

eba99 committed:
Improvements to Dash bulkuploader

Automatic refresh of api instance to avoid query timeouts.
Refine get_load_info function

eba99 committed:
Remove default dtype = object when reading files for upload

Causes all columns to be varchar in the lake, which may be undesirable.

eba99 committed:
Remove test for invalid load type

This is handled in lowerlevel load api class, and so tests should be performed there.

eba99 committed:
Revert "Remove test for invalid load type"

This reverts commit 5e4e9dd1b2b06a53b6bf5ec60eae0b5710e86cad.

eba99 committed:
Test that lowerlevel load is called without checking parameters for now

Better validation built into the function, so ignore this until issue can be resolved.

eba99 committed:
Add tests for upload_df in Load class

eba99 committed:
Fix chunking dtype issue

Providing a dtype based on first chunk to avoid potential commit issues down the line

eba99 committed:
Improve validation on upload of file and query

eba99 committed:
Mock refresh api instance in query tests

eba99 committed:
Update unit tests for DashBulkUploader

eba99 committed:
Add patch for remove_load test

eba99 committed:
Add dtype fixing code to upload query function

eba99 committed:
Update to require python version 3.9

eba99 committed:
Update to python version 3.9

eba99 committed:
Update dash.py

eba99 committed:
Allow to specify dtype in bulk uploader

eba99 committed:
Update Pipfile.lock

eba99 committed:
Update index.rst

eba99 committed:
Update index.rst

eba99 committed:
Fix heading lengths

eba99 committed:
Finalize documentation

timalive committed:
Update conf.py

timalive committed:
Merge branch 'add_utility_upload_function' of https://github.com/ComotionLabs/comotion-sdk into add_utility_upload_function

eba99 commented:
The following has been changed in this branch:
- The read_and_upload_file_to_dash function was re-written to support both the v1 and v2 data model.  This is to allow for smoother transition for clients after migrating the data model
- The Load class was enhanced to allow easy uploads from a pandas dataframe, various file types, a directory and a dash.Query object
- A DashBulkUploader class was created to allow for easy uploading to multiple lake tables at once.  This uses the load class and uses a dictionary to manage several loads with multi-threading implemented to improve upload performance.

So essentially the old upload function has been modified to be backwards-compatible, in a way.  The read_and_upload_file_to_dash function uses the v1 api if not migrated, and the v2 load mechanism via the DashBulkUploader otherwsie.  The DashBulkUploader uses the Load class, which will be the core, public-facing class to upload data to Dash.  So everything flows through the load class eventually anyway to ensure upload behaviour is consistent.

The SDK documentation was also updated substantially, to add new examples to the overview and refresh the documentation for the new classes created in the dash.py source file.

eba99 committed:
Merge auth.py from main

eba99 committed:
Merge test_auth.py from main

eba99 committed:
Update Pipfile.lock

eba99 committed:
Merge branch 'main' into add_utility_upload_function

eba99 committed:
Update Pipfile.lock

eba99 committed:
Add back exception group to pipfile

eba99 committed:
Add error to remove load/data when load has been committed

Also add tomli for pytest

eba99 committed:
Add numpy to dev packages

eba99 committed:
Add numpy to packages

eba99 committed:
Update unit tests for updated remove_load and remove_data_from_load

eba99 committed:
Update pipfile

eba99 committed:
Add pandas to dev packages

timalive committed:
update dependencies

timalive reviewed (changes requested):
see my comments on review.

eba99 committed:
Remove comotion-sdk from packages and add unit tests

Add tests for: upload_file, upload_query, remove data from load where load is not 'OPEN'

eba99 committed:
Fix out-of-sync pipfile.lock

eba99 committed:
Update test_integration_tests.py

eba99 committed:
Add deprecation warning and refresh docs

eba99 committed:
Add ability to use file name as file key in upload file

Also wrote unit tests for scenarios

timalive committed:
remove file_upload_response as an option from upload_df

eba99 committed:
Add explanation of entity type
